File: cloakbrowser/widevine.py
"""Widevine CDM hint-file seeding for persistent contexts.

CloakBrowser's binary is built with Widevine support but ships no CDM (the CDM
is a proprietary Google binary we can't redistribute). Users sideload it by
copying a ``WidevineCdm/`` directory from a real Chrome install next to the
binary (see issue #96).

Chromium discovers a sideloaded CDM in two phases: an early-startup pass that
reads a "hint file" from the user-data-dir, and a later async component-updater
pass that writes that hint file. On a fresh profile the hint file doesn't exist
on the first launch, and Playwright passes ``--disable-component-update``, so the
updater never writes it — Widevine only works after a manual two-launch dance.

This module pre-seeds the hint file before launch so a sideloaded CDM works on
the very first launch. It never bundles, downloads, or copies the CDM itself —
it only writes the hint when a CDM the user provided is already present.

Linux only: Chromium's hint-file mechanism is Linux/ChromeOS-specific. On Windows
the CDM can't initialise (DRM host verification), and macOS uses a different CDM
layout, so seeding is a no-op there.
"""

from __future__ import annotations

import json
import logging
import os
import platform
from pathlib import Path

logger = logging.getLogger("cloakbrowser")

# Chromium reads this file from <user-data-dir>/WidevineCdm/ at early startup.
_HINT_FILENAME = "latest-component-updated-widevine-cdm"


def _seeding_disabled() -> bool:
    """True if CLOAKBROWSER_WIDEVINE is set to a falsey value (kill switch)."""
    val = os.environ.get("CLOAKBROWSER_WIDEVINE", "").strip().lower()
    return val in ("0", "false", "off", "no")


def resolve_widevine_cdm_dir(binary_path: str | os.PathLike) -> Path | None:
    """Locate a sideloaded Widevine CDM directory, or None if absent.

    Resolution order:
      1. If CLOAKBROWSER_WIDEVINE_CDM is set, it is used **exclusively** (overrides
         auto-detection). An invalid value (no ``manifest.json``) skips seeding.
      2. ``<dir of the chrome binary>/WidevineCdm`` — where a user naturally drops
         a manual sideload, per Chromium binary version.
      3. ``<cache dir>/WidevineCdm`` (``~/.cloakbrowser/WidevineCdm``) — the
         version-independent location the Docker auto-fetch and ``fetch-widevine.py``
         write to. This fallback lets one fetched CDM serve any binary (free or
         Pro, any version) with no env var — the CDM ``.so`` is arch-specific but
         not version-specific.

    A directory counts only if it contains ``manifest.json`` (so we don't seed a
    hint pointing at a bogus path). The returned path is absolute and
    symlink-resolved (``Path.resolve()``).
    """
    custom = os.environ.get("CLOAKBROWSER_WIDEVINE_CDM")
    if custom is not None:
        # Set exclusively (overrides auto-detection). An empty/whitespace value is
        # invalid — return None rather than let Path("") resolve to "." and match a
        # stray manifest.json in the working directory.
        if not custom.strip():
            return None
        cdm_dir = Path(custom)
        return cdm_dir.resolve() if (cdm_dir / "manifest.json").is_file() else None

    from .config import get_cache_dir  # local import avoids any import-cycle risk

    for cdm_dir in (Path(os.fspath(binary_path)).parent / "WidevineCdm",
                    get_cache_dir() / "WidevineCdm"):
        if (cdm_dir / "manifest.json").is_file():
            return cdm_dir.resolve()
    return None


def seed_widevine_hint(user_data_dir: str | os.PathLike, binary_path: str | os.PathLike) -> None:
    """Write the Widevine CDM hint file into a persistent profile before launch.

    ``binary_path`` is the resolved chrome executable; the CDM is looked for next
    to it. No-op on non-Linux platforms, when seeding is disabled via
    CLOAKBROWSER_WIDEVINE, or when no sideloaded CDM is present. Never raises —
    a failure here must not break the browser launch.
    """
    if platform.system() != "Linux":
        return
    if _seeding_disabled():
        logger.debug("Widevine hint seeding disabled via CLOAKBROWSER_WIDEVINE")
        return
    if not user_data_dir:
        # Empty user_data_dir = Playwright's ephemeral profile (its own temp dir);
        # a persistent hint can't be placed there, and "" would pollute the CWD.
        return

    # Everything below is best-effort and must never break the browser launch,
    # so the whole body (resolution + write) is guarded.
    try:
        cdm_dir = resolve_widevine_cdm_dir(binary_path)
        if cdm_dir is None:
            if os.environ.get("CLOAKBROWSER_WIDEVINE_CDM") is not None:
                logger.warning(
                    "CLOAKBROWSER_WIDEVINE_CDM is set but has no manifest.json; "
                    "skipping Widevine hint seeding"
                )
            else:
                logger.debug("No sideloaded Widevine CDM found; skipping hint seeding")
            return

        hint_dir = Path(os.fspath(user_data_dir)) / "WidevineCdm"
        hint_dir.mkdir(parents=True, exist_ok=True)
        hint_file = hint_dir / _HINT_FILENAME
        # cdm_dir is already absolute/resolved. Compact separators + ensure_ascii=False
        # byte-match the JS wrapper's JSON.stringify (UTF-8) output.
        content = json.dumps({"Path": str(cdm_dir)}, separators=(",", ":"), ensure_ascii=False)

        try:
            if hint_file.is_file() and hint_file.read_text(encoding="utf-8") == content:
                return  # already seeded correctly
        except Exception:
            logger.warning("Existing Widevine hint unreadable; rewriting")

        hint_file.write_text(content, encoding="utf-8")
        logger.info("Seeded Widevine CDM hint -> %s", cdm_dir)
    except Exception as e:
        logger.warning("Failed to seed Widevine CDM hint file: %s", e)

File: js/src/widevine.ts
/**
 * Widevine CDM hint-file seeding for persistent contexts.
 * Mirrors Python cloakbrowser/widevine.py.
 *
 * CloakBrowser's binary supports Widevine but ships no CDM (proprietary, can't
 * redistribute). Users sideload it by copying a `WidevineCdm/` directory from a
 * real Chrome install next to the binary (see issue #96). Chromium reads a
 * "hint file" from the user-data-dir at early startup to register the CDM, but
 * on a fresh profile it doesn't exist yet, and Playwright disables the component
 * updater that would write it. This seeds the hint file before launch so a
 * sideloaded CDM works on the first run. It never bundles, downloads, or copies
 * the CDM — only writes the hint when a user-provided CDM is already present.
 *
 * Linux only: Chromium's hint-file mechanism is Linux/ChromeOS-specific.
 */

import fs from "node:fs";
import path from "node:path";

import { getCacheDir } from "./config.js";

const HINT_FILENAME = "latest-component-updated-widevine-cdm";

/** True if `file` exists and is a regular file (mirrors Python's Path.is_file()). */
function isFile(file: string): boolean {
  try {
    return fs.statSync(file).isFile();
  } catch {
    return false;
  }
}

/** Absolute, symlink-resolved path (mirrors Python's Path.resolve()). */
function realPath(p: string): string {
  try {
    return fs.realpathSync(p);
  } catch {
    return path.resolve(p);
  }
}

function seedingDisabled(): boolean {
  const val = (process.env.CLOAKBROWSER_WIDEVINE ?? "").trim().toLowerCase();
  return val === "0" || val === "false" || val === "off" || val === "no";
}

/**
 * Locate a sideloaded Widevine CDM directory, or null if absent.
 *
 * Resolution order:
 *   1. If CLOAKBROWSER_WIDEVINE_CDM is set, it is used exclusively (overrides
 *      auto-detection). An invalid value (no `manifest.json`) skips seeding.
 *   2. `<dir of the chrome binary>/WidevineCdm` — a manual sideload, per version.
 *   3. `<cache dir>/WidevineCdm` (`~/.cloakbrowser/WidevineCdm`) — the
 *      version-independent location the Docker auto-fetch and fetch-widevine.py
 *      write to. This fallback lets one fetched CDM serve any binary (free or
 *      Pro, any version) with no env var — the CDM `.so` is arch- not version-specific.
 *
 * A directory counts only if it contains `manifest.json`. The returned path is
 * absolute and symlink-resolved (mirrors Python's Path.resolve()).
 * @internal Exported for testing.
 */
export function resolveWidevineCdmDir(binaryPath: string): string | null {
  const custom = process.env.CLOAKBROWSER_WIDEVINE_CDM;
  if (custom !== undefined) {
    // Set exclusively (overrides auto-detection). An empty/whitespace value is
    // invalid — return null rather than let path.join("", ...) match a stray
    // manifest.json in the working directory.
    if (custom.trim() === "") return null;
    return isFile(path.join(custom, "manifest.json")) ? realPath(custom) : null;
  }
  for (const cdmDir of [
    path.join(path.dirname(binaryPath), "WidevineCdm"),
    path.join(getCacheDir(), "WidevineCdm"),
  ]) {
    if (isFile(path.join(cdmDir, "manifest.json"))) return realPath(cdmDir);
  }
  return null;
}

/**
 * Write the Widevine CDM hint file into a persistent profile before launch.
 * `binaryPath` is the resolved chrome executable; the CDM is looked for next to
 * it. No-op on non-Linux, when disabled via CLOAKBROWSER_WIDEVINE, or when no
 * sideloaded CDM is present. Never throws — a failure must not break launch.
 */
export function seedWidevineHint(userDataDir: string, binaryPath: string): void {
  if (process.platform !== "linux") return;
  if (seedingDisabled()) return;
  // Empty userDataDir = Playwright's ephemeral profile (its own temp dir);
  // a persistent hint can't be placed there, and "" would pollute the CWD.
  if (!userDataDir) return;

  // Everything below is best-effort and must never break the browser launch,
  // so the whole body (resolution + write) is guarded.
  try {
    const cdmDir = resolveWidevineCdmDir(binaryPath);
    if (cdmDir === null) {
      if (process.env.CLOAKBROWSER_WIDEVINE_CDM !== undefined) {
        console.warn(
          "[cloakbrowser] CLOAKBROWSER_WIDEVINE_CDM is set but has no manifest.json; " +
            "skipping Widevine hint seeding",
        );
      }
      return;
    }

    const hintDir = path.join(userDataDir, "WidevineCdm");
    fs.mkdirSync(hintDir, { recursive: true });
    const hintFile = path.join(hintDir, HINT_FILENAME);
    // cdmDir is already absolute/resolved.
    const content = JSON.stringify({ Path: cdmDir });

    try {
      if (isFile(hintFile) && fs.readFileSync(hintFile, "utf-8") === content) {
        return; // already seeded correctly
      }
    } catch {
      console.warn("[cloakbrowser] Existing Widevine hint unreadable; rewriting");
    }
    fs.writeFileSync(hintFile, content);
  } catch (e) {
    // Best-effort: never break the launch, but surface the failure.
    console.warn("[cloakbrowser] Failed to seed Widevine CDM hint file:", e);
  }
}
